### ISSUE TITLE
Improve help system of kdb

### ISSUE THREAD

**Issue by markus2330** [ElektraInitiative/libelektra#260]
--help typically is not useful for someone who is new in Elektra.

Improve built-in information, add more markdown pages to "kdb info".

Expected outcome: At least Linux-Gurus should be able to use kdb without any external help.

**Comment by iandonnelly** [comment]
Alright, so I looked into this a bit and got my dev environment back up and running. I do have a few questions though:

1) What do you envision the help text to look like? Is there an example you can give me about another program with a really good help text in your opinion?

2) For the `kdb info` command, do you know specific plugins that are missing any information or do you just want more information for some plugins? 

3) Also for `kdb info`, is using the markdown from the GitHub pages the best solution for this, or should we have shorter blurbs about each plugin here? In other words should these be in-depth explanations of the plugins or just summaries of what they do?

**Comment by markus2330** [comment]
1) I think the docu of git is done really nicely. @mhaberler or others: do you have other suggestions?

2.) Some plugins should contain easier introductions about what the plugins are doing and how this influences the behaviour. E.g. the resolver should give some easy to digest information about where configuration is stored and how you can find it (`kdb file`).

I also think about adding other Markdown documents to "kdb info", which are not directly implemented by a plugin, e.g. "kdb info mounting" and "kdb info plugins" should give general information about the mounting and plugins concepts.

To resolve the issue its enough if you just write the markdown documents, I can do the integration so that they are displayed when "kdb info" is executed.

3.) At some point (reachable by links) there should be an in-depth explanation. Obviously we do not want to duplicate the same text, so there should be pages with general information and specific plugins link to that information.

E.g. `kdb mount --help` should mention `kdb info plugins`

**Comment by mhaberler** [comment]
Please - just write a _man page_ which reflects what this thing can do, and how - before going into a _internal help system_ development effort.

**Comment by markus2330** [comment]
Actually, my plan is to convert the markdown pages (and --help text) to man pages. Directly writing man pages would be more effort and would result to nearly unreadable pages in github.

**Comment by iandonnelly** [comment]
As I am going through the commands I just want to say that having an `ls` command to list keys and a `list` command to list plugins is very confusing. I suggest renaming the `list` command to `plugins` as it seems easier and more intuitive. Maybe we can have a hidden `list` command that is a virtual version of plugin to prevent breaking things for a few versions?

**Comment by markus2330** [comment]
You are completely right!

I would suggest to use a hierarchical naming schema:
kdb-plugin (no real command, just general info about plugins)
kdb-plugin-check
kdb-plugin-info
kdb-plugin-list

To be discussed if this also applies to meta and/or mounting:

kdb-meta (info about meta data)
kdb-meta-ls
kdb-meta-set

kdb-backend-mount
kdb-backend-unmount
kdb-backend-file (to resolve an Elektra path to a file system path)

and for future global backends:

kdb-gbackend-mount
kdb-gbackend-unmount
kdb-gbackend-printfile

I agree we should have aliases to not break scripts.

We should also hide some less important commands from --help (and give them to list-tools), e.g. fstab, test, vset is hardly used.

**Comment by markus2330** [comment]
Btw. I suggest that we use [ronn-format(7)](https://rtomayko.github.io/ronn/ronn.7.html).

It is markdown with some the extensions, e.g.
1.) description lists (I was always missing that!)
2.) the first line contains special information to generate the header/footer of manpages.

The current drawback is that (quote from BUGS in ronn(1)):

> Ronn is written in Ruby and depends on hpricot and rdiscount, extension libraries that are non-trivial to install on some systems. A more portable version of this program would be welcome.

But ruby-ronn is trivial to install on debian and the possibility to build man-pages is not essential (we can the build-server let them build).

So what do you say?

**Comment by markus2330** [comment]
(Btw. it looks nice with github markdown compiler, too)

**Comment by iandonnelly** [comment]
It looks pretty good to me, I was wondering about the best way to convert them. Also you are talking about definition lists? Where you can define arguments or whatnot, I think that's an important feature. Also what are the man pages going to consist of exactly? Or should I just worry about the help texts for now? Also re-writing the help texts isn't the quickest process. I have them all as markdown pages right now but still need to edit them. I am going to do a simple one and a more complicated one first and show you how they look to make sure I am going in the right direction. Should be done with those tomorrow.

**Comment by iandonnelly** [comment]
Also, it is pretty trivial to install on Debian and I don't think it'll be a real issue since only developers of Elektra would need this package (and even then not all of them).

**Comment by markus2330** [comment]
In the end the man pages should contain all usual sections. But yes, as first step, you should only worry about the help text + how to use the programs with some examples.

The SYNOPSIS and ENVIRONMENT are subject to be generated, so it would be useless to write them now.

Its a good idea if you do a preliminary merge request first. Then I can start implementing the man page generation in parallel.

**Comment by iandonnelly** [comment]
A few things.

1. There seems to be some bug with ronn where you can't nestle lists in definition lists, see my [merge documentation](https://github.com/iandonnelly/iansdonnelly-libelektra/blob/help-documentation/doc/help/MERGE.md) and how ronn handles it. I can't think of a good work-around right now, any suggestions?

2. My branch is a little messy for a merge request right now because I have all the help texts saved in doc/help/* but most are unedited still.

3. What's up with the (nickel plugin test)[src/plugins/ni/nickel-1.1.0/src/tests/ni_parse_spaces_quotes.ini] which keeps showing up in my commits? I see you had an issue with it changing when running cmake, what's the best way to handle it?

**Comment by markus2330** [comment]
Ian Donnelly wrote:
> 1. There seems to be some bug with ronn where you can't nestle lists in
> definition lists, see my [merge
> documentation](https://github.com/iandonnelly/iansdonnelly-libelektra/blob/
> help-documentation/doc/help/MERGE.md) and how ronn handles it. I can't think
> of a good work-around right now, any suggestions?

I think this should be avoided. Best, you simply refer to `[STRATEGIES]
(STRATEGIES)` and have a own section explaining the strategies (again in 
definition lists).

Note that the OPTIONS will be generated, otherwise its hard to keep it 
consistent. Avoid copying existing help text and modify it somewhere else.
OPTIONS text can be modified here:

https://github.com/iandonnelly/iansdonnelly-libelektra/blob/help-documentation/src/tools/kdb/cmdline.cpp


> 2. My branch is a little messy for a merge request right now because I have
> all the help texts saved in doc/help/* but most are unedited still.

You can create a new branch and just copy what is finished for merging.

> 3. What's up with the (nickel plugin
> test)[src/plugins/ni/nickel-1.1.0/src/tests/ni_parse_spaces_quotes.ini]
> which keeps showing up in my commits? I see you had an issue with it
> changing when running cmake, what's the best way to handle it? 

I think thats only a issue of discouraged in-source tree builds? You can 
simply put that file in .git/exclude.

**Comment by iandonnelly** [comment]
Just want to let you know I was traveling all day yesterday and busy today. It's going to be a slow week for me here until I get home Wednesday but I'll definitely put in some time for the fixes you already commented on and maybe a few of the simpler commands.

**Comment by markus2330** [comment]
@iandonnelly Considering the good quality you are delievering: I think its better if you fully write the man page as it should be. Maybe we will remove the internal help system. And even if not, your man pages will help us to determine what is the correct final man page.

So you can add SYNOPSIS, the `--help` output and so on.

It is also the direction as suggested by @mhaberler.

Is this better for you this way?

**Comment by markus2330** [comment]
I think we finally made this one! A huge thanks for all involved, you will be mentioned in the release notes in the first section ;)

**Comment by iandonnelly** [comment]
Awesome! I have started training for a new job so I have been super busy right now but I'm glad all of this is done now and the new docs will be in the new release!
